### DISCUSSION TITLE
M2M logical AND composition

### DISCUSSION THREAD

**Discussion by ethagnawl** [MrThearMan/graphene-django-extensions#91] [Q&A] [answered]
I am in the process of migrating an application I'm working on to this library. We're excited about using many of its features, but the one that provides an immense amount of immediate value is the filter functionality. This allows us to "simply" do complex logical operations which we were previously doing in one-off filter functions and were not readily composable.

The one problem which I have not yet found a straightforward/scalable* answer to is how to how we can dynamically compose Q objects which do a logical AND query across associated fields (e.g. `Foo.objects.filter(tags__name="season 1")`).

I am working on shimming a functional example into the example_app included in this repository but, in the meantime, here is the fundamental distinction (depth vs. breadth) between what the GDX filter logic produces and what is required to fit this use case:

filter Qs generated by UserDefinedFilter#build_user_defined_filters:
```
(AND: ('tags__name__exact', 'season 1'), ('tags__name__exact', 'episode 1'))
<FooQuerySet []>
```

Iteratively applied filters (i.e. winnowing):
```
queryset = Foo.objects.all()
queryset = queryset.filter(tags__name="season 1")
queryset = queryset.filter(tags__name="episode 1")
<FooQuerySet [<Foo: SOME-UUID>]>
```

\* The one escape hatch that we can potentially fall back on is first excluding Foos whose tag__names we are not interested. This does not scale well, though, and adds an undue burden to the client.

Assuming this is not currently possible using the exposed logical operators, I was thinking that it might be possible to extend the defined filter operations and introduce logic that builds up a list of winnowing filter operations (e.g. `thunks = [lambda x: x.filter(tags__name="season 1")]`) and then applies them as part of an overridden UserDefinedFilter#build_user_defined_filters method but it's not clear if/how they could be made composable and to work with the existing operators/workflow.

For reference, this association is introduced by the Taggit library and [this issue](https://github.com/jazzband/django-taggit/issues/680) lays out the problem very nicely.

**Comment by ethagnawl** [comment]
Here's a branch which modifies the example_app to provide a reproducible test bed for the scenario outlined above: https://github.com/ethagnawl/graphene-django-extensions/pull/1

**Comment by ethagnawl** [comment]
I have been able to achieve rudimentary support for OR and AND using the lambda approach outlined above using the following modifications:

```
    def filter(self, qs: QuerySet, data: UserDefinedFilterInput) -> QuerySet:
        if data in EMPTY_VALUES:
            return qs

        result = self.build_user_defined_filters(data)

        if result.annotations:  # pragma: no cover
            qs = qs.annotate(**result.annotations)
        if result.ordering:  # pragma: no cover
            qs = qs.order_by(*qs.query.order_by, *result.ordering)

        for filter_thunk in result.filter_thunks:
            qs = filter_thunk['OP'](qs)

        qs = qs.distinct()

        return qs

    def build_user_defined_filters(self, data: UserDefinedFilterInput) -> UserDefinedFilterResult:
        filters: Q = Q()
        filter_thunks = []
        ann: dict[str, Any] = {}
        ordering: list[str] = []

        if data.operation.value in ("AND", "OR", "XOR", "NOT"):
            if data.operations is None:
                msg = "Logical filter operation requires 'operations' to be set."
                raise ValueError(msg)

            if data.operation.value == "NOT" and len(data.operations) != 1:
                msg = "Logical filter operation 'NOT' requires exactly one operation."
                raise ValueError(msg)

            for operation in data.operations:
                result = self.build_user_defined_filters(operation)
                # if result.annotations:  # pragma: no cover
                #     ann.update(result.annotations)
                # if result.ordering:  # pragma: no cover
                #     ordering.extend(result.ordering)

                if data.operation.value == "AND":
                    xs = []
                    for filter_thunk in result.filter_thunks:
                        filter_thunk["_OP"] = "AND"
                        filter_thunk["OP"] = filter_thunk["AND"]
                        xs.append(filter_thunk)
                    filter_thunks = filter_thunks + xs

                elif data.operation.value == "OR":
                    xs = []
                    for filter_thunk in result.filter_thunks:
                        filter_thunk["_OP"] = "OR"
                        filter_thunk["OP"] = filter_thunk["OR"]
                        xs.append(filter_thunk)
                    filter_thunks = filter_thunks + xs
                elif data.operation.value == "NOT":
                    raise NotImplementedError("NOT")
                    # filters = ~result.filters
                elif data.operation.value == "XOR":
                    raise NotImplementedError("XOR")
                    # filters ^= result.filters

        else:
            if data.field is None:
                msg = "Comparison filter operation requires 'field' to be set."
                raise ValueError(msg)

            alias: str = getattr(data.field, "name", data.field)
            field: str = self.extra["fields"][alias]
            inputs: dict[str, Any] = {f"{field}{LOOKUP_SEP}{data.operation.value.lower()}": data.value}

            for k, v in inputs.items():
                # TODO: Is items() necessary?
                arg = {}
                arg[k] = v
                ops = {
                    "OP": lambda xx: xx,
                    "AND": lambda xx: xx.filter(**arg),
                    "OR": lambda xx: xx | xx.filter(**arg),
                }
                print('arg', arg)
                print('ops', ops)
                filter_thunks.append(ops)

        return UserDefinedFilterResult(
            filters=filters,
            filter_thunks=filter_thunks,
            annotations=ann,
            ordering=ordering
        )
```

This is experimental at best and I'm not by any means proposing that this ever make its way upstream. I also need to see about reimplementing NOT and XOR, validate that order_by and annotate still work, ensure that the Queryset is still evaluated lazily/as late as possible, factor out some of the common behavior, etc.

**Comment by MrThearMan** [comment]
Hi, sorry for the late response!

Ah, the classic "[lookups spanning multi-valued relationships](https://docs.djangoproject.com/en/5.1/topics/db/queries/#spanning-multi-valued-relationships)" problem. Indeed, this is not possible with just `Q()` objects, due to SQL details.

<details>

<summary>Open this for a deeper explanation.</summary>



---

Let's say we have a table `Example` with a many-to-many relationship to a table `Tag`. Both tables have the same two columns: `id` and `name`.

We have 2 rows in our `Example` table:

| Example ID | Example name |
| --- | --- |
| 1 | fizz |
| 2 | buzz |

We also have 3 rows in the `Tag` table

| Tag ID | Tag name |
| --- | --- |
| 1 | foo|
| 2 | bar |
| 3 | baz |

Suppose Tag 1 and Tag 2 are connected to Example 1 and also that Tag 1 and Tag 3 are connected to Example 2 (both have two tags).

| Example ID | Example name | First tag name | Second tag name |
| --- | --- | --- | --- |
| 1 | fizz | foo | bar |
| 2 | buzz | foo | baz |

Now, what we want to fetch from this data are the rows in the `Example` table that have relations to both the "foo" and "bar" rows in the `Tag` table. To do this in django, we need to use `example_qs.filter(tags__name='foo').filter(tags__name='bar')`, which boils down to the following SQL.

```sql
SELECT
    e.id,
    e.name
FROM example e
INNER JOIN example_tag et1 ON   e.id     = et1.example_id
INNER JOIN tag          t1 ON et1.tag_id =  t1.id
INNER JOIN example_tag et2 ON   e.id     = et2.example_id
INNER JOIN tag          t2 ON et2.tag_id =  t2.id
WHERE (
        t1.name = 'foo'
    AND t2.name = 'bar'
);
```

Notice that we need to make 4 joins to accomplish this. It makes more sence if we look at the joined table without the where clause, and put in the tag names in the `SELECT` as well.

| e.id | e.name | t1.name | t2.name |
| --- | --- | --- | --- |
| 1 | fizz | foo | foo |
| 1 | fizz | foo | bar |
| 1 | fizz | bar | foo |
| 1 | fizz | bar | bar |
| 2 | buzz | foo | foo |
| 2 | buzz | foo | baz |
| 2 | buzz | baz | foo |
| 2 | buzz | baz | baz |

Essentially, we are creating all combinations of joining these tables together. From this, it's easier to see that if we want the rows where `t1.name = 'foo'` and `t2.name = 'bar'`, there is only one option, which is the second row.

While it may be possible to do this in other ways, this is by far the easiest. Still it does require us to join the `Tag` table and the `ExampleTag` table both `N` number of times, where `N`=number of tags we want Example row to have. Something to keep in mind.

---

</details>

Essentially, what we are asking here is "Give me all rows in table `Foo` that have all of these tags: `episode_1` and `season_1`". Therefore, I think the API should be something like this:

```graphql
query {
 foos(
    filter: {
      field: tags
      operation: ALL
      operations: [
        {
          field: name
          operation: EXACT
          value: "foo"
        },
        {
          field: name
          operation: EXACT
          value: "bar"
        },
      ]
    }
  ) {
    id
    name
  }
}
```

In the filter, we would need to validate that the `ALL` operation is only used then the field is a `many-to-many` or `one-to-many` relation. As for the algorithm, I think `UserDefinedFilterResult.filters` might need to be a list that contains `Q()` objects that should be added in separate `queryset.filter` calls. I don't think we need to use lambdas if we do it like that.

I might have a crack at this this weekend, though you are welcome to share your thoughts and concerns before that.

**Reply by ethagnawl** [reply]
Thanks for following up with such a detailed summation of the problem and a sound proposal for a solution, @MrThearMan. Also, you've got nothing to apologize for -- quite the opposite in fact!

Your proposal sounds like a very nice one and more elegant or, at least, much less intrusive than mine. I wish I had more thoughts to add but it sounds like you've got a good handle on it all. We will be more than happy to beta test for you, if/when you get around to implementing this.

Thanks again! :100:

**Comment by MrThearMan** [comment]
Hi, I was able to make this change quite easily, and it's now available in [v0.4.6](https://github.com/MrThearMan/graphene-django-extensions/releases/tag/v0.4.6)

**Comment by ethagnawl** [comment]
Thanks, @MrThearMan! We're very excited about opting into this new feature! :pray:

**Comment by MrThearMan** [comment, marked as answer]
@ethagnawl Hey, I noticed there was a bug in the implementation of [v0.4.6](https://github.com/MrThearMan/graphene-django-extensions/releases/tag/v0.4.6). Fixing this changes the API a little bit:

```graphql
query {
 foos(
    filter: {
      operation: ALL
      operations: [
        {
          field: tagName
          operation: EXACT
          value: "foo"
        },
        {
          field: tagName
          operation: EXACT
          value: "bar"
        },
      ]
    }
  ) {
    id
    name
  }
}
```

You would also need to define "tagName" on the filter explicitly like this:

```python
    filter = UserDefinedFilter(
        model=....,
        fields=[
            ...
            ("tags__name", "tag_name"),
        ],
    )
```

This is in line with how related fields are handled at the moment (you need to define all of them), since there is no "infinite model intropection" built into the `UserDefinedFilter` (not sure if this is even possible).

Anyway, this is fixed in [0.4.7](https://github.com/MrThearMan/graphene-django-extensions/releases/tag/v0.4.7). Cheers!

**Reply by ethagnawl** [reply]
Thanks, once again, @MrThearMan!
